File: app/src/main/java/com/example/mike/mp3player/service/library/content/searcher/SongSearcher.java
package com.example.mike.mp3player.service.library.content.searcher;

import android.content.ContentResolver;
import android.database.Cursor;
import android.provider.BaseColumns;
import android.provider.MediaStore;

import com.example.mike.mp3player.commons.MediaItemType;
import com.example.mike.mp3player.service.library.content.parser.ResultsParser;
import com.example.mike.mp3player.service.library.search.Song;
import com.example.mike.mp3player.service.library.search.SongDao;

import org.apache.commons.lang3.StringUtils;

import java.util.ArrayList;
import java.util.List;

import static com.example.mike.mp3player.service.library.content.Projections.SONG_PROJECTION;

public class SongSearcher extends ContentResolverSearcher {

    private static final String PARAMETER = "?";

    public SongSearcher(ContentResolver contentResolver, ResultsParser resultsParser, String idPrefix, SongDao songDao) {
        super(contentResolver, resultsParser, null, idPrefix, songDao);
    }

    @Override
    String[] getProjection() {
        return SONG_PROJECTION.toArray(new String[0]);
    }

    @Override
    public MediaItemType getSearchCategory() {
        return MediaItemType.SONGS;
    }

    @Override
    public Cursor performSearchQuery(String query) {
        List<Song> results =  searchDatabase.query(query);
        List<String> ids = new ArrayList<>();
        List<String> parameters = new ArrayList<>();
        if (results != null && !results.isEmpty()) {
            for (Song song : results) {
                ids.add(song.getId());
                parameters.add(PARAMETER);
            }

        }


        String WHERE_CLAUSE = BaseColumns._ID + " IN(" + StringUtils.join(parameters, ", ") + ") COLLATE NOCASE";
        String[] WHERE_ARGS = ids.toArray(new String[ids.size()]);
        return  contentResolver.query(MediaStore.Audio.Media.EXTERNAL_CONTENT_URI,
                getProjection(),
                WHERE_CLAUSE, WHERE_ARGS, null);
    }
}

File: service/src/main/java/com/github/goldy1992/mp3player/service/library/content/searcher/SongSearcher.kt
package com.github.goldy1992.mp3player.service.library.content.searcher

import android.content.ContentResolver
import android.database.Cursor
import android.provider.BaseColumns
import android.provider.MediaStore
import com.github.goldy1992.mp3player.commons.MediaItemType
import com.github.goldy1992.mp3player.service.library.MediaItemTypeIds
import com.github.goldy1992.mp3player.service.library.content.Projections
import com.github.goldy1992.mp3player.service.library.content.parser.SongResultsParser
import com.github.goldy1992.mp3player.service.library.data.search.Song
import com.github.goldy1992.mp3player.service.library.data.search.SongDao
import kotlinx.coroutines.CoroutineScope
import org.apache.commons.lang3.StringUtils

open class SongSearcher

    constructor(contentResolver: ContentResolver,
                resultsParser: SongResultsParser,
                private val mediaItemTypeIds: MediaItemTypeIds,
                songDao: SongDao,
                scope: CoroutineScope)
    : ContentResolverSearcher<Song>(
        contentResolver,
        resultsParser,
        null,
        songDao,
        scope) {
    override val idPrefix: String
        get() = mediaItemTypeIds.getId(MediaItemType.SONG)

    override val projection: Array<String?>
        get() = Projections.SONG_PROJECTION.toTypedArray()

    override val searchCategory: MediaItemType?
        get() = MediaItemType.SONGS

    override suspend fun performSearchQuery(query: String?): Cursor? {
        val results: List<Song>? = searchDatabase.query(query)
        val ids: MutableList<String> = ArrayList()
        val parameters: MutableList<String?> = ArrayList()
        if (results != null && !results.isEmpty()) {
            for (song in results) {
                ids.add(song.id)
                parameters.add(PARAMETER)
            }
        }
        val WHERE_CLAUSE = BaseColumns._ID + " IN(" + StringUtils.join(parameters, ", ") + ") COLLATE NOCASE"
        val WHERE_ARGS = ids.toTypedArray()
        return contentResolver.query(MediaStore.Audio.Media.EXTERNAL_CONTENT_URI,
                projection,
                WHERE_CLAUSE, WHERE_ARGS, null)
    }

    companion object {
        const val PARAMETER = "?"
    }

}